### PR TITLE
port over focusable-arrow-keys-behavior

### PR DESCRIPTION
**Pull request by jstefaniuk-d2l**
Ports over this:
https://github.com/Brightspace/d2l-polymer-behaviors-ui/blob/6614114c98c2578ef44b56132216bbaa07955260/d2l-focusable-arrowkeys-behavior.js

Which will likely be used in a few core components

### CODE REVIEW

**Comment by jstefaniuk-d2l** [comment]
Whoops I guess fastdom is being removed from things: https://github.com/BrightspaceUI/core/commit/9792fd330ef578216965ffdb03cb4d7722fafa73

I better not add it back XD

**Comment by dbatiste** [comment]
Thanks @jstefaniuk-d2l  for taking this on.  This is definitely something that we'll want to use in a couple of other components.  

Per our conversation, I think we'll want to make this a `mixin`.  I think the bulk of the logic can remain identical, but we'll want to change the API slightly to follow the best practice for Lit-element's `UI = f(data)`, where `f = render`.  Essentially, we should not be modifying the DOM outside of `render` (which includes not wiring up event handlers that provide behavior).  In this case, the event handler should be wired up as part of `render` using Lit-element's `@` event directive.  These are the key differences (see code mock-up below):

* container is defined by mixin template instead of `arrowKeyFocusablesContainer`
* consumer indicates focusable elements with `d2l-arrowkeys-focusable` classes.  The mixin query for these elements instead of calling a provider.  We can introduce the provider API later if we need more flexibility. 

Mixin:
```
export const ArrowKeysMixin = superclass => class extends superclass {

   static get properties() {
      return {
         arrowKeysDirection: { type: String },
         arrowKeysNoWrap: { type: Boolean }
       };
   }

   constructor() {
      super();
      this.arrowKeysDirection = 'leftright';
      this.arrowKeysNoWrap = false;
   }

   arrowKeysContainer(inner) {
      return html`<div @keyup="${this._handleArrowKeys}">
         ${inner}
      </div>`;
   }

   _handleArrowKeys() {
      // arrow keys logic
   }

   // ... the bulk of the existing logic

};
```

Consumer:

```
class MyElement extends ArrowKeysMixin(LitElement) {

   constructor() {
      super();
      this.arrowKeysDirection = ...;  // optionally override default
      this.arrowKeysNoWrap = ...;  // optionally override default
   }

   render() {
      const inner = html`
         <a href="..." class="d2l-arrowkeys-focusable">link 1</a>
         <a href="..." class="d2l-arrowkeys-focusable">link 2</a>
         // etc
      `;
      return html`
         <div class="some-outer-html">
            ...
            ${this.arrowKeysContainer(inner)}
            ...
         </div>
      `;
   }

   async arrowKeysBeforeFocus() {
      // consumer optionally implements
   }

}
customElements.define('my-element', MyElement);
```
Of course, the consumer can compose whatever HTML as needed.

One other subtle difference here is wire-up to `keyup` instead of `keydown`.  I think that is what we would want for a11y.  Hopefully that doesn't cause any regression.

**Comment by jstefaniuk-d2l** [comment]
TODO:
- [ ] Documentation file

**Comment by dbatiste** [comment]
Looking good.

I think we'll want some unit tests for this as well.  For the most part, the existing tests can be copied over from the [existing tests](https://github.com/Brightspace/d2l-polymer-behaviors-ui/blob/master/test/focusable-arrowkeys-behavior.html) and [test component](https://github.com/Brightspace/d2l-polymer-behaviors-ui/blob/master/test/focusable-arrowkeys-behavior-components.js) as a Lit-element.  I think we'll want to follow [what we've been doing in helpers](https://github.com/BrightspaceUI/core/tree/master/helpers/test) for that.  I'd probably just copy one of them and replace titles, fixtures, and tests.

`/mixins/test/arrow-keys-mixin.html` (tests)
`/mixins/test/arrow-keys-mixin-components.js` (test-component that uses the mixin)

File: mixins/arrow-keys-mixin.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# ArrowKeysMixin
+
+Used for managing focus with the arrow keys.
```
Comment (dbatiste) [inline mixins/arrow-keys-mixin.md]: Perhaps we can reformat a little so the code can fit into view, and use more than one sample to show the usage?    

> The focusable elements can be provided in one of two ways.
>
> If the elements are known up-front and are in the element's local DOM scope, simply add the `d2l-arrowkeys-focusable` class.
```javascript
import { ArrowKeysMixin } from '@brightspace-ui/core/mixins/arrowkeys-mixin.js';
class MyElement extends ArrowKeysMixin(LitElement) {
   render() {
      return html`
         <div>
            ...
            <a href="..." class="d2l-arrowkeys-focusable">link 1</a>
            <a href="..." class="d2l-arrowkeys-focusable">link 2</a>
            ...
         </div>
      `;
   }
}
customElements.define('my-element', MyElement);
```
> If the elements are not known up front, or the elements cannot be simply queried in the element's local DOM-scope, an `async arrowKeysFocusablesProvider` may be implemented.
```javascript
import { ArrowKeysMixin } from '@brightspace-ui/core/mixins/arrowkeys-mixin.js';
class MyElement extends ArrowKeysMixin(LitElement) {
   render() {
      return html`
         <div>
            ...
            <a href="...">link 1</a>
            <a href="...">link 2</a>
            ...
         </div>
      `;
   }
   async arrowKeysFocusablesProvider() {
      return [ /* array containing focusable elements */]
   }
}
customElements.define('my-element', MyElement);
```
Lastly, we can include a properties section that lists the optional properties.

***Properties:***

* `arrowKeysDirection` (optional): The RTL behavior (default is `leftright`)
* `arrowKeysNoWrap` (optional): To opt out of wrapping focus from end-to-start and start-to-end
* `arrowKeysBeforeFocus` (optional): Async callback invoked before focus us applied

File: mixins/arrow-keys-mixin.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+		this.arrowKeysNoWrap = false;
+	}
+
+	arrowKeysContainer(inner) {
+		return html`<div class="arrow-keys-container" @keyup="${this._handleArrowKeys}">
+			${inner}
+		</div>`;
+	}
+
+	async focusablesProvider() {
```
Comment (dbatiste) [inline mixins/arrow-keys-mixin.js]: I think this should be `arrowKeysFocusablesProvider` to be consistent with the rest of the API.

File: mixins/arrow-keys-mixin.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	arrowKeysContainer(inner) {
+		return html`<div class="arrow-keys-container" @keyup="${this._handleArrowKeys}">
+			${inner}
+		</div>`;
+	}
+
+	async arrowKeysFocusablesProvider() {
+		return [...this.shadowRoot.querySelectorAll('.d2l-arrowkeys-focusable')];
```
Comment (dbatiste) [inline mixins/arrow-keys-mixin.js]: Is this being done to convert to an array?  If so, it this to work around some browser inconsistency?

File: mixins/test/localize-mixin.html [hunk 4]
```diff
@@ -9,18 +9,18 @@
 		<script src="/node_modules/chai/chai.js"></script>
 		<script src="/node_modules/@polymer/test-fixture/test-fixture.js"></script>
 		<script src="/node_modules/wct-mocha/wct-mocha.js"></script>
-		<script type="module" src="./test-elem.js"></script>
+		<script type="module" src="./localize-test-elem.js"></script>
```
Comment (dbatiste) [inline mixins/test/localize-mixin.html]: Awesome - thanks for updating this!

File: mixins/test/arrow-keys-mixin.html [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	<meta charset="utf-8">
+	<meta name="viewport" content="width=device-width, minimum-scale=1, initial-scale=1, user-scalable=yes">
+	<title>ArrowKeysMixin test</title>
+	<script src="/node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
+	<script src="/node_modules/mocha/mocha.js"></script>
+	<script src="/node_modules/chai/chai.js"></script>
+	<script src="/node_modules/@polymer/test-fixture/test-fixture.js"></script>
+	<script src="/node_modules/wct-mocha/wct-mocha.js"></script>
+	<script type="module" src="./arrow-keys-test-elem.js"></script>
+	<script type="module" src="../arrow-keys-mixin.js"></script>
```
Comment (dbatiste) [inline mixins/test/arrow-keys-mixin.html]: Is the mixin file needed here?

File: mixins/test/localize-mixin.html [hunk 6]
```diff
@@ -9,18 +9,18 @@
...
 		<script type="module" src="../localize-mixin.js"></script>
```
Comment (dbatiste) [inline mixins/test/localize-mixin.html]: Seems like this mixin file is also not needed.

**Pull request merged**